File: CS333 - Programming Languages/Semantics/Go/task2.go
/*
 * File: task2.go
 * Purpose: To show function variables works in Go
 * Author: Samuel Munoz
 * Date: 03-07-2020
 */

package main
import ( 
	"fmt"
	"os"
	"strconv"
)

func add(a, b int) int { return a+b }
func subtract(a, b, c int, calc func(int, int) int) int { return a-calc(b,c) }

func main() {
	if len(os.Args) < 3 {
		fmt.Printf("command: ./task2 <integer> <integer> <integer>")
	} else {
		var calc func(int, int) int
		calc = add
		a, _ := strconv.Atoi(os.Args[1])
		b, _ := strconv.Atoi(os.Args[2])
		c, _ := strconv.Atoi(os.Args[3])
		fmt.Printf("Sum: %d\n", calc(a, b))
		fmt.Printf("Difference: %d\n", subtract(a, b, c, calc))
	}
}

File: CS333 - Programming Languages/Semantics/C++/task2.cpp
/*
 * File: task2.cpp
 * Purpose: To show how I can make variables with data type function
 * Author: Samuel Munoz
 * Date: 03-07-2020
 */

#include <iostream>
#include <cstdlib>

int add(int a, int b) {
	return a+b;
}

int subtract(int a, int b, int c, int (*calc)(int, int)) {
	return a-calc(b,c);
}

int main(int args, char **argv) {
	if(args < 4) {
		printf("command: ./task2 <integer> <integer> <integer>\n");
		return 0;
	}
	
	int (*calc)(int, int);
	calc = add;

	int v1 = atoi(argv[1]);
	int v2 = atoi(argv[2]);
	int v3 = atoi(argv[3]);
	std::cout << "Add value: " << calc(v1, v2) << std::endl;

	std::cout << "Difference value: " << subtract(v1, v2, v3, calc);

	return 0;
}

/*
Source
https://www.cprogramming.com/tutorial/function-pointers.html
https://codeforwin.org/2017/12/pass-function-pointer-as-parameter-another-function-c.html
*/